### ISSUE TITLE
A new internal fast-lookup data structure for target information

### ISSUE THREAD

**Issue by wlandau** [ropensci/drake#440]
### Lessons from #435 and #283:

- `drake` spends a lot of time repeatedly looking up target-level information it should already know.
- The dependency network is inflexible.

### Explanation

I think these and similar problems come from the limitations of `drake`'s data structures. The `plan` is great for users, but lookup is slow, and it contains only indirect information about each target's dependencies. The `graph` has complete dependency information, but it does not have other metadata (`igraph` attributes are a pain), and traversing the whole graph for dependencies is slow.

### Solution

I want to put all target-level metadata in a unified fast-lookup data structure. This new `config$meta` will replace individual outputs of `drake_meta()` flying around, and it will decrease reliance on the `plan` and `graph`. For [fast lookup](https://stackoverflow.com/questions/34232865/r-fast-single-item-lookup-from-list-vs-data-table-vs-hash), I want to base this data structure on an environment. For convenience, I want it to have a `storr`-like API. Best of both worlds: a [`storr_environment()` cache](http://richfitz.github.io/storr/reference/storr_environment.html). (Thank you @richfitz for making it available!)

There is a lot of refactoring to do, but I think this will make `drake` a whole lot faster and give it a cleaner code base.

**Comment by wlandau** [comment]
For the sake of #283, we should probably distinguish between the build schedule and the constituent files and objects. Maybe we need separate `storr_environment()`s:

1. `config$meta`: a `storr_environment()` for individual data objects. Should have the following `storr` namespaces:
    - `name`: character string
    - `imported`: logical, imported from the user's workspace/environment or generated with a `drake` command?
    - `file`: whether the item is a file or simply a generic R object.
    - `missing`: whether the item is present in `config$envir`
    - `hash`: the object's fingerprint.
    - `mtime`: file modification time (`NULL` for non-files)
2. `config$protocol`:
    - `"name"`: name of the step in the schedule, usually a target or import
    - `"command"`: command from the `drake_plan()` (`NULL` for imports)
    - `"seed"`: RNG seed
    - `"trigger"`: the trigger from the `drake_plan()` (or the default trigger)
    - `"object_in"`: names of non-file dependencies
    - `"file_out"`: names of any file outputs (potentially more than one)
    - `"file_in"`: names of any non-`knitr` input files
    - `"knitr_in"`: names of any `knitr` input files
    - `"dependency_hash"`: A hash used to trip the `depends` trigger.
    - `"file_hash"`: A hash used to trip the `file` trigger.

**Comment by wlandau** [comment]
We should also maintain overall `all_targets` and `all_imports` lists in `config`.

**Comment by kendonB** [comment]
Forgive me if this is a dumb question, but what's the barrier to just using the plan `data.frame` for everything and just using R's standard  `[` lookups or `dplyr::filter`? These aren't nearly as fast as a hash table but they don't require building the table, which you would have to do for every make, I presume. I couldn't seem to find any benchmarks comparing `storr_environment` to something simple like `filter` but maybe something is out there.

ref for benchmarking `[`, `filter` and the `data.table` equivalents: https://appsilondatascience.com/blog/rstats/2017/03/02/r-fast-lookup.html

**Comment by wlandau** [comment]
There is no barrier, technically. I am glad you checked, especially in https://github.com/richfitz/storr/issues/81. From [these results](https://stackoverflow.com/questions/34232865/r-fast-single-item-lookup-from-list-vs-data-table-vs-hash), plain environments seem to win out in terms of lookup speed, and I just assumed `storr_environment()`s would offer similar speed and more convenience.

From https://github.com/richfitz/storr/issues/81#issuecomment-401631474, I now think the right approach is to have `config$meta` be a plain `new.env()` with all the individual `drake_meta()` lists (augmented with dependency information). Lookup speed, code style, and my ability to solve #435 and #283 will all improve.

**Comment by wlandau** [comment]
Edit: existing `drake_meta()` info needs to be computed at the last minute, while dependency information can be resolved all at once up front. I'm thinking this lookup structure can just focus on being a more efficient version of `config$graph`, possibly even replacing `config$graph` at some point.

**Comment by wlandau** [comment]
This data structure is really a protocol for building the targets. Whereas the `drake_plan()` tries to be friendly to the user, the protocol tries to be friendly to the internals.

**Comment by wlandau** [comment]
To do: put an `R6` wrapper around `config$protocol` with a bunch of safe accessor methods (would also help enforce `inherit = FALSE` in `get()`).

**Comment by kendonB** [comment]
Is it worth benchmarking `get` wrapped up in the R6 piece before going that way?

**Comment by wlandau** [comment]
Should be easy enough to check.

**Comment by wlandau** [comment]
I am going to put this issue on hold because

1. #435 was solved a different way. With `igraph_options(return.vs.es = FALSE)`, performance issues with graph lookups may not be such a big deal.
2. In my work on #457, I realized that #440 may complicate the internals more than it is worth. It is extra work and clutter to maintain a new data structure, and it adds confusion because `config$protocol` conceptually overlaps with the roles of `config$plan`, `config$graph`, and `drake_meta()`.
2. #369 is all that could stand to gain from #440 right now, and I want to solve it in a way that cleans up the code instead of complicating it.

**Comment by wlandau** [comment]
Then again, this issue is so huge in scope that more dust needs to settle before I am likely to attempt it seriously.

**Comment by wlandau** [comment]
Also, we really do need to think about situations where a graph is really handy: for example, `prune_envir()` with the `"lookahead"` strategy and an efficient `outdated()` look all the way downstream on the graph. Can the lookup object and priority queue really do the same thing?

**Comment by wlandau** [comment]
From #483, the complete dependency information of targets is neatly arranged in `igraph` attributes. With those attributes organized, I think we can revisit a lookup structure to cut back on the multitude of linear-time lookups in `drake`. Steps I will be taking to approach the problem:

-[ ] Deprecate `build_drake_graph()` and replace it with `drake_config(...)$graph`.
-[ ] Divide the old `build_drake_graph()` internals into `create_drake_lookup()` and `create_drake_graph()`, the latter of which will use the lookup structure to create the graph. Neither of these new functions will be exported. I will not give special attributes to the `igraph` object.
-[ ] In addition to the original `igraph` attributes, the new lookup structure will have the target-specific elements of the plan. No more linear-time subsetting, e.g. `plan$command[plan$target == "my_target]`. This operation in particular will be replaced by the constant-time `config$lookup[["my_target"]]$command`.

**Comment by wlandau** [comment]
I plan to encapsulate the lookup operation internally.

```r
look_up <- function (target, field, config){
  info <- get(
    x = target,
    envir = config$lookup,
    inherit = FALSE
  )
  get(
    x = field,
    envir = info,
    inherit = FALSE
  )
}
```

Here, `info` is just a pointer (an environment) I doubt the first lookup will put much burden on the L1-L3 caches.

We can think about managing internal metadata this way instead of passing around a `meta` object, but that's an aesthetic matter we can come back to later. I actually kind of like the separation between `drake_meta()` objects and `config$lookup`. The former is information light enough to store as persistent metadata in `drake`'s cache, and the important bits changes when the target is built. The latter has permanent target-level information that we don't really need to put in storage.

**Comment by wlandau** [comment]
But to prepare for a more complete transition to this lookup data structure, maybe we should call it `config$meta` (to keep open the possibility of consolidating `drake_meta()` objects) and call the `look_up()` function `get_meta()` instead. We could also pair it with a `set_meta()` function.

**Comment by wlandau** [comment]
Before diving headlong into this refactoring, I want to do more in-depth profiling. The `magrittr` pipe tends to complicate `profvis` output, so a first step is to completely clean out `%>%`.

**Comment by wlandau** [comment]
Thanks to #571, `profvis` now delivers useful results. #572 accounts for half the target-level overhead we see in `drake_example("overhead")`. Given the amount of work and care it would take to solve #440, I am postponing it again until it becomes an actual bottleneck.

**Comment by wlandau** [comment]
See #575.

**Comment by wlandau** [comment]
This feature may not reduce overhead, but it should simplify the architecture.

**Comment by wlandau** [comment]
Just started the work [here](https://github.com/ropensci/drake/tree/440).

**Comment by wlandau** [comment]
New favorite term for this structure: "layout". I also thought about "topology", but it seems more redundant with "graph" and implies stronger-than-necessary mathematical claims.

**Comment by wlandau** [comment]
Fixed via #576.

**Comment by wlandau** [comment]
On reflection, the term "layout" is not quite right. Let's go with "workflow specification", or "spec" for short. The spec is `drake`'s interpretation of the plan. The `drake` plan is user-friendly, and all the dependency relationships are *implicit*. The spec is machine-friendly, and all the dependency relationships are *explicit*. `drake_config()` takes the plan and makes the spec by doing static code analysis on the commands and functions (i.e. `analyze_code()`).

The name change is implemented in 322785ead30be24da59ed8f19fbebb26c43c3803.
